### DISCUSSION TITLE
Configuration and configuration file

### DISCUSSION THREAD

**Discussion by royfalk** [vegastrike/Vega-Strike-Engine-Source#912] [General]
Moved from https://github.com/vegastrike/Assets-Production/pull/135#issuecomment-2495944357

--------------------

@evertvorster :
"Since this is a new venture in settings for Vega Strike, can we agree on a rough list of requirements for the configuration app?

Having the settings app separate from the main application is weird, and counter-intuitive.

Ideally, the game engine should call the settings app on startup.
The settings app should then check to see if there is a valid configuration file.
If there is, check to see that the requested resolution and screen in the configuration file is available.
If it is, and the settings app was just called to check that screen is available for the game to run on, pass that information back to the game engine, together with the requested screen information.

If the configuration file does not exist, or does not have a valid screen configuration anymore, give the user the choice of which screen and resolution to run on, and save that to the config file, then return that information back to the game engine.

Now, you should also be able to call the configuration app from the game menu. When you call it like this, then the configuration app should expose all the configuration options and at the end write that to the configuration file, and return back to the game.

It is my humble opinion that this should be in place before we start porting over each of the settings that is possible with the old settings app. We could then incrementally move over settings that makes sense. There are also quite a few settings in the original configuration file that is not settable in the app, and you needed to modify the configuration file directly. Stuff like assigning keys to buttons on the joystick, which axes does what and so on. That might be a good reach goal for sometime in the future."

**Comment by royfalk** [comment]
> Having the settings app separate from the main application is weird, and counter-intuitive.

I debated this (with myself). Here are the main reasons why I chose to place it in assets:
1. I plan to implement it in python. The engine repo doesn't have a good place for it.
2. We may want different settings apps for different games. Different background, colors, fonts but most important, different settings. WC doesn't have ftl or SPEC. No need then for these to exist.

> Ideally, the game engine should call the settings app on startup.
...

Not sure I agree. This complicates the current system a lot. 
Right now the game relies on a configuration file to start.Without it, the game will crash with an error.
Ideally, the game should simply fall back to defaults for each setting instead. 

> Now, you should also be able to call the configuration app from the game menu. 

I agree. But this is a lot more complicated for now because the game's GUI is pretty much hard coded. It also means changing the in-memory config as well as the file and handling any other edge cases. For example, changing resolution should change the game resolution in game.

> It is my humble opinion that this should be in place before we start porting over each of the settings that is possible with the old settings app. 

It wasn't my plan to touch the settings app at all. IIRC, I started with the support for selecting monitors. However, now that I did, I plan to quickly write a replacement and ditch it. The main reason is that the old app is horribly out of date and confusing:

1. _Computer_ sets a whole bunch of other settings.
2. Geometry is probably obsolete and we could all play on 'Extreme Detail'.
3. Same for textures - unlimited.
4. Same for faction textures - "more vram". 
5. Same for shaders.
6. Resolution field should be updated with modern resolutions and the minimum should be full-HD or so.
7. Color is always 32bit.
8. Accelerated visual - always on.

**Reply by BenjamenMeyer** [reply]
Keeping this part of the discussion here:

> > It is my humble opinion that this should be in place before we start porting over each of the settings that is possible with the old settings app.
> 
> It wasn't my plan to touch the settings app at all. IIRC, I started with the support for selecting monitors. However, now that I did, I plan to quickly write a replacement and ditch it. The main reason is that the old app is horribly out of date and confusing:
> 
> 1. _Computer_ sets a whole bunch of other settings.
> 2. Geometry is probably obsolete and we could all play on 'Extreme Detail'.
> 3. Same for textures - unlimited.
> 4. Same for faction textures - "more vram".
> 5. Same for shaders.
> 8. Accelerated visual - always on.

I've actually still found these useful on some of the systems I plan on; not all of them are able to go to the top for performance. That said, we could probably adjust how we do it and certainly adjust defaults. But we should also keep in mind that not everyone is playing on a brand new, always upgraded gaming rig.

> 6. Resolution field should be updated with modern resolutions and the minimum should be full-HD or so.

This should be informed by querying SDL2 instead of just some random set that is hard-coded.
Also, we should default to a Windows Display, especially as that's what is most useful for debugging. (And yes, I pretty much only play in a windowed display, not full screen.)

> 7. Color is always 32bit.

Agreed.

**Comment by evertvorster** [comment]
My reasoning for the game engine calling the settings app is a little along the following lines:

The settings app needs to have code that checks screen resolutions, and what is available. It also needs code to read the config file and hopefully do some error detection and so on. 

Rather than implementing that in the engine again, just have the engine ask the settings app for a valid configuration to play from. 

So, less duplicating of code. Maybe the vsettings app can have a --check mode that just returns true if the config file is good?
The reason for this is that as monitors are plugged in and out, the display might not be possible anymore, and rather than crashing the game can deal with this situation gracefully.

**Reply by BenjamenMeyer** [reply]
so I don't think it needs to call the settings app; but it does need to validate the settings are valid in some form.
This might be through an option related to the settings app; but it could easily be done outside of the settings app too.

So ultimately:
- yes, the engine should validate the settings on start-up
- this involves settings that are both related to the engine and a secondary set of settings specific to the game assets.
- if the settings are invalid, then the game should default to some settings (e.g resolution) to get a working UI and then automatically load the settings configuration, ideally highlighting the invalid settings.

**Comment by evertvorster** [comment]
There are a lot of obsolete options in the settings app, I agree. 
Also, the ability to set font size would have been really nice. 
In most of the game setting a larger font pitch in vegastrike.config works really well. It is only in the initial menus that seem to have a fixed vertical spacing in pixels and so a larger font overlaps the text on the next line.

**Reply by BenjamenMeyer** [reply]
Font size will likely need to wait until we can handle that better in the code; but yes we should plan for it. We could put something in now we a value to show it's coming and then just disable it until we can actually do something about it.

**Comment by evertvorster** [comment]
As per your request, the config.json file needs to be in a place where the user can modify it without system privileges. 
In a similar idea to how the vegastrike.config file is treated, the vegastrike.config file that is distributed with Assets is seen as a default configuration file, and if no other copy of this file exists it is used as the game configuration. 
The vsettings app has a similar mechanism, where it looks for a user-modified vegastrike.config file. If it fails to find that, it uses the default configuration, and then when the user saves the configuration it is saved in ~/.vegastrike on Linux.

Something similar should exist for config.json to enable changes made to the default configuration to survive when the Assets are refreshed with an upgrade.

**Reply by BenjamenMeyer** [reply]
So a good pattern for configuration is:

- defaults (typically stored with the game installation)
- system config (e.g /etc, %ProgramDataDir%)
- user config (${HOME}, user profile)

The system should start by loading game defaults, then override with system values, then override again with user values.

Now,  for VS the system values are probably not very important so possible we could skip that step; but we should have at least the two steps with the game defaults being something reasonable.

**Comment by BenjamenMeyer** [comment]
> > Having the settings app separate from the main application is weird, and counter-intuitive.
> 
> I debated this (with myself). Here are the main reasons why I chose to place it in assets:
> 
> 1. I plan to implement it in python. The engine repo doesn't have a good place for it.
> 2. We may want different settings apps for different games. Different background, colors, fonts but most important, different settings. WC doesn't have ftl or SPEC. No need then for these to exist.

There should be settings exposed from the game engine, and then contributed to by the game assets.

For the engine itself, we should have a configuration object - libconfig - that deals with all the universal game engine configuration values. This would get exposed to the Python side so that the Game Assets can provide a UI for viewing and setting them. Some can be allowed to be read-only if we don't want them to be set but the game assets need to use them.

I do agree that the Game Assets should own the Settings interface so that it can be skinned to match the game.

**Comment by BenjamenMeyer** [comment]
>> Now, you should also be able to call the configuration app from the game menu.

> I agree. But this is a lot more complicated for now because the game's GUI is pretty much hard coded. It also means changing the in-memory config as well as the file and handling any other edge cases. For example, changing resolution should change the game resolution in game.

So the game should be able to present the access to the settings menu from the different points:
- main menu
- Base Computer (when landed)
- Ship Computer (when in flight)

Granted the UI stuff is pretty hard-coded right now so let's focus on just getting it to the main menu first with the idea that it'll show up other places in the future, and we can work towards making things more flexible in the process as a long term goal.

**Comment by stephengtuggy** [comment]
@royfalk have you seen https://github.com/protocolbuffers/protobuf ? It looks like it's written to solve largely the same problem set that you're looking to solve here.

Conceivably, we could write our settings spec in `.proto` files, then generate code to work with it in C++, Python, and other languages.

**Reply by royfalk** [reply]
It solved a similar but different problem. I wanted the C++ object to be generated directly from the json, automatically load the whole json and make it into a singleton. Of the three requirements, the only one supported is possibly the loading.

What I wrote is more specialized for a very specific reason - we have a very large config file. As we convert to JSON and modify the config schema, I wanted this to be reflected in C++ without any coding on my part. It took me about two three hours to write and can now take most of the scut work out of a long process.

I didn't bother putting this project as part of VS as it isn't. Once we migrate the config file. I don't expect many changes and it's not needed. If you add a new setting or two, you can do it manually without breaking anything.

**Reply by stephengtuggy** [reply]
AFAIK, protobuf would solve the first two requirements, just maybe not the third (singleton). Code would be auto-generated from .proto files, which can express richer information about data types and such than plain JSON can.

Whatever solution we end up with, we need to make sure we can handle all the different setting types:

- Sections and subsections
- Plain text settings
- Integer settings
- Floating-point settings, some of them double-precision
- Currency amounts
- Color values
- Boolean values
- File names and paths
- Screen resolution, aspect ratio; base, computer, and ship resolution and aspect ratio
- Music tracks
- Points (locations) in 3D space
- Ships, missions, starting positions?

... and possibly others that I'm forgetting.

**Reply by stephengtuggy** [reply]
Also, please let's be careful when changing the order of operations in the startup code for the game engine. It's easy to mess up -- and even easier to mess up on a platform other than one's own, without realizing it.

**Reply by stephengtuggy** [reply]
Then there is the hierarchy of which directory locations to look in for the settings files, mission files, etc. That is a complicated tangle that probably deserves a discussion of its own.

**Reply by BenjamenMeyer** [reply]
So I really don't like protobuf. Too much autogenerated code among other things. it's not something we need and would overly complicate the code to support it

**Reply by royfalk** [reply]
I think you missed something about my goals. I wanted to **only** edit the JSON file and have the C++ be auto-generated. Creating a protobuf schema and JSON file is twice the work. vegastrike.config is 1500 lines. 

I am aware that by not including a schema (JSON or protobuf), we are losing the ability to enforce correctness. However, my solution isn't the best. It's merely good enough.

As for the order of operations and directory locations, I've been moving away from it and using standard boost and C++ file operations. Assuming root folder is Asset-Production, that should be good enough.

**Reply by BenjamenMeyer** [reply]
@royfalk I'm generally good with the direction you've been going. Just noting my objection to protobuf.

**Reply by stephengtuggy** [reply]
OK. protobuf was just an idea. We don't have to go that route.

**Comment by evertvorster** [comment]
Thinking about this a little more.. From a very high level:

Because the default configurations are in system space, and not to be modified, and .vegastrike/vegastrike.config is very old and we would like to move away from it, here is a suggestion:

Let's call the settings binary on every game start. 
The setting binary would contain code to read the old config file, and the new one, and do detections of screens and available resolutions and other features that don't really belong in the game. 
When settings starts, it checks for the config files in a pre-determined user-modifyable location. On Linux this would be ~/.vegastrike/
If the file(s) are found, then check version and whether the screen and resolution defined in the config file are available. 
If it is all good, don't even pop up the UI, just return control to the game engine with a no errors detected signal. 

If the config file is not found, or there is something wrong with it, then settings brings up the UI to adjust settings and ultimately either save the settings file, or quit.

Settings should be able to detect screens and resolutions available. Commonly set options should also be read from a system file, or detected. There should be an advanced tab where you can adjust fuel burn rates and that sort of thing. 

Once the configuration is done, settings saves the config file, and hands over control to the game binary again, stating that the settings files are good, and can be used.

The advantages of doing this is that it is automatic, the config file location is controlled by us, and so we don't need a whole lot of places to look for it, and it can be versioned. 

The reason for versioning the config file is that as options change, we would like to update the config file. 


Of course, we would like all the configuration to be done in one file only, but to achieve this there will have to be a transition period where the game reads variables from both new and old configuration files, and so we need a settings app that can also read & write to both configuration files. 

Once we don't use the old config file anywhere in the game engine anymore, support for it can be dropped from the settings app. 

Are there any obvious flaws in the logic here? Or maybe a better way to do it?

**Reply by royfalk** [reply]
Generally, I think we all agree with you on most stuff here (especially high level). 
Implementation wise, it gets more messy:

- There's an undergoing effort to port the old config to the new. Some of the discussion here (e.g. protobuf) is about how to do it in the most efficient way. This is not a technically complicated issue. The file is simply very big and hard to port in one go.
- The new file is currently just in Assets. We'd like a mechanism to store changes to it in .vegastrike. This requires development.
- The new files work fine together. They are both loaded into the same struct but to different keys.
- Screen and resolution auto detection needs development. 
- In game settings app needs development.
- Versioning of the JSON config file could get messy. Especially if we break the format somehow. One issue is different versions. You'll have the same config version for both master and 10.0-beta-3. maybe append version or branch name to config file and have multiple files??

**Reply by BenjamenMeyer** [reply]
I'd like to do away with the separate settings binary which would make this flow easier - if the settings are not valid then we automatically load the settings configuration with a predetermined configuration that would work.

Keep in mind that the `~/.vegastrike` directory is dependent on the Assets used.

**Reply by BenjamenMeyer** [reply]
@royfalk we'd want to encode a format version into the file too, and then have a way of upgrading it to the latest used by the engine that loaded it.

**Comment by evertvorster** [comment]
To add a thought in this thread... 

Since what is written into the config files are evolving, the game engine should check if a variable exists in the user config file. 

If it does not, then the game engine should write the default value of that variable into the user's config file. 
Else the game engine should use the variable in the config file. 


As for hierarchy of save files causing confusion, on my system there are only two. The default one that is installed by the game, and the one created by the user when the program is run for the first time. 
Variables in the latter should take precedence over values in the former, if they exist.

**Reply by BenjamenMeyer** [reply]
If the value doesn't exist, then a log message should be generated about it and the compile time default value should be used. 
When the user initiates a config write then the game can update and write those values down; but it shouldn't happen automatically.

**Reply by BenjamenMeyer** [reply]
We could add a command-line parameter to tell the game engine it's allowed to write/update the config upon read.
Perhaps this could be available in several methods:

| Option(s) | Description |
|---|---|
| --config-validate | check the configuration to ensure it's valid |
| --config-update --dry-run | report what would be updated if the config were to be written, but do NOT write it |
| --config-update | report what would be updated if the config where to be written AND write it |

**Reply by evertvorster** [reply]
Using a default value for a variable that should be in the config file and not actually writing that value into the config file seems to be obscuring the fact that such a configurable variable exists. 

Really, what is wrong with writing the default value of a variable into a config file if that variable did not exist in the config file?

Let's take an example to illustrate my point:
Lets say vs gains a config value to set the color of the computer in bases. 
The engine knows the default value for this, but since it is new, it is not mentioned in the config file. 

Now, if the engine did not write the value into the config file, it would just be using the default value forever, and the user would not know that this is even a value that can be changed. 

If the engine did write the default value to the config file, nothing would change for the user, as the default value will still be used. However, once the default is changed in the config, then the value in the config file gets used. 

Of course, when the variable is in the configuration file, the engine would never modify that variable, it would just use what is there.

**Reply by BenjamenMeyer** [reply]
It's a user experience thing. We shouldn't just overwrite configs without the user telling us to. We can certainly make a big deal about it in the logs and make a recommendation to enable rewriting the configs.

**Comment by vincele** [comment]
I think we should not overwrite user's config without at least asking him if it's OK to do so.
Or we should make backups (without ever overwriting old ones) beforehand.

I did not re-read the whole thread, maybe this point has already been discussed, if so, please disregard my comment

**Reply by stephengtuggy** [reply]
@vincele I think we are on the same page there. The only settings that @evertvorster was talking about adding automatically are the new ones, that the user hasn't set yet.

Making backups also seems like a good idea.

**Reply by BenjamenMeyer** [reply]
@vincele agreed we shouldn't just write it automatically; however, it should happen the next time the config gets written down.

**Comment by stephengtuggy** [comment]
So in https://github.com/vegastrike/Vega-Strike-Engine-Source/pull/1078 , I implemented the two config files, one in Assets-Production and one in ~/.vegastrike, with settings from the latter overwriting settings from the former. I just closed #1078 since it's too large. But implementing that mechanism wasn't too difficult. It should be easy to implement again in a smaller PR.
